compiler/rustc_next_trait_solver/src/solve/mod.rs

```diff
@@ -148,6 +148,45 @@ where
         }
     }
 
+    fn compute_unstable_feature_goal(
+        &mut self,
+        param_env: <I as Interner>::ParamEnv,
+        symbol: <I as Interner>::Symbol,
+    ) -> QueryResult<I> {
+        // Iterate through all goals in param_env to find the one that has the same symbol.
+        for pred in param_env.caller_bounds().iter() {
+            if let ty::ClauseKind::UnstableFeature(sym) = pred.kind().skip_binder() {
+                if sym == symbol {
+                    return self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes);
+                }
+            }
+        }
+
+        // During codegen we must assume that all feature bounds hold as we may be
+        // monomorphizing a body from an upstream crate which had an unstable feature
+        // enabled that we do not.
+        //
+        // The coherence tests in
+        // tests/ui/unstable-feature_bound/unstable_impl_coherence.rs
+        // will fail if we return error instead of ambiguity.
```

Comment:
```suggestion
        // Coherence should already report overlap errors involving unstable impls
        // as the affected code would otherwise break when stabilizing this feature.
        // It is also easily possible to accidentally cause unsoundness this way as
        // we have to always enable unstable impls during codegen.
```